src/core/tests/widgets/test_canvas.py

```diff
@@ -12,3 +13,105 @@ def setUp(self):
     def test_widget_created(self):
         self.assertEqual(self.testing_canvas._impl.interface, self.testing_canvas)
         self.assertActionPerformed(self.testing_canvas, 'create Canvas')
+
+    def test_basic_drawing(self):
+        with self.testing_canvas.save_restore():
+            self.testing_canvas.rect(-3, -3, 6, 6)
+            self.testing_canvas.fill_style('rgba(0, 0.5, 0, 0.4)')
+            self.testing_canvas.fill(preserve=True)
+            self.testing_canvas.stroke_style('rgba(0.25, 0.25, 0.25, 0.6)')
+            self.testing_canvas.line_width(1)
+            self.testing_canvas.stroke()
```

Comment:
These tests validate that using the interface API doesn't raise errors, but it doesn't validate that the underlying API is actually invoked. The changes I added to the testing framework yesterday should allow you to validate the sequence of events that have occurred. You might need to add another assertion to the base TestCase; but validating, for example, that this test results in a context save, a series of draw operations, then a context restore would be helpful.

Reply:
I added a test_canvas_on_draw method to try to better cover these cases. I haven't yet added a complete sequence of events test with a new assertion yet.